operator/v1/types_ingress.go

```diff
@@ -1032,6 +1032,53 @@ type IngressControllerHTTPHeaderBuffer struct {
 	HeaderBufferMaxRewriteBytes int32 `json:"headerBufferMaxRewriteBytes,omitempty"`
 }
 
+// HSTSDirectiveType is an optional directive that directs HSTS Policy configuration.
+//+kubebuilder:validation:Enum=preload;includeSubDomains
+type HSTSDirectiveType string
+
+const (
+	// PreloadHSTSDirective directs the client to include hosts in its host preload list so that
+	// it never needs to do an initial load to get the HSTS header (note that this is not defined
+	// in RFC 6797 and is therefore client implementation-dependent).
+	PreloadHSTSDirective HSTSDirectiveType = "preload"
+
+	// IncludeSubDomainsHSTSDirective means the HSTS Policy applies to any subdomains of the host's
+	// domain name, when the route admission policy allows wildcards.  If wildcard routes aren't allowed,
+	// this directive has no effect.
+	IncludeSubDomainsHSTSDirective HSTSDirectiveType = "includeSubDomains"
+
+	AllHSTSHeaderPolicyScope     = "All"
+	LimitedHSTSHeaderPolicyScope = "Limited"
+)
+
+type IngressControllerHSTSHeaderPolicy struct {
+	// maxAgeSeconds is the delta time range in seconds during which hosts are regarded as HSTS hosts.
+	// If set to 0, it negates the effect and hosts are no longer regarded as HSTS hosts.
+	// maxAgeSeconds is a time-to-live value, and if this policy is not refreshed on a client, the HSTS
+	// policy will eventually expire on that client.
+	// +required
+	MaxAgeSeconds int32 `json:"maxAgeSeconds"`
+
+	// HSTSDirectives is a list of optional HSTS directives that direct HSTS Policy configuration.
+	// +kubebuilder:validation:UniqueItems=true
+	// +optional
+	HSTSDirectives []HSTSDirectiveType `json:"hstsDirectives,omitempty"`
+
+	// scope indicates whether this header policy is enabled for a limited set of routes that have hosts
+	// with domains in the list specified in inScopeDomains (limited scope) or for all TLS routes
+	// (all scope).  If HSTS is desired for all routes, set the scope field to "All". If HSTS is desired
+	// for only a limited set of routes, set the scope field to "Limited" and add a list of subdomains
+	// to the InScopeDomains field.  If scope is omitted, no HSTS header policy will be added.
+	// +optional
+	Scope string `json:"scope,omitempty"`
```

Comment:
Should scope also be an enum?

Reply:
I need to be able to pass in `scope` from the HAProxy template, and decided to keep it as a string rather than implement an enum in the template.

Reply:
> I need to be able to pass in scope from the HAProxy template

Could you elaborate on what you mean by this? I can't think of a scenario where information for an API field would be passed in from the HAProxy template, since typically API fields are consumed by the HAProxy template in some manner.

I'm wondering if the `Scope` field is even necessary: Could we assume a scope of `All` if `InScopeDomains` is empty?

Reply:
There is a scenario described in the enhancement proposal where the scope is retrieved from the environment and passed into the `hstsHeaderValue` helper function: https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-options-enable-hsts.md#enable-hsts-per-domain

We debated a default scope of `All` in the ehancement proposal and decided against it.

Reply:
Ah, right, I do recall not wanting `All` scope behavior to be the default, so having the `Scope` field would make sense.
It's still not clear to me why `Scope` can't be an enum value. The ingress-operator could cast the enum into a string when it sets the router deployment's environment variables.

Reply:
The template calls `hstsHeaderValue` helper function directly via a go template like this:
`  {{- with $hsts := hstsHeaderValue $router_global_hsts $router_hsts_scope $cfg $router_hsts_list }}
` 
`hstsHeaderValue` is defined in pkg/router/template/template_helper.go as one of the HAProxy template `helperFunctions`.

The $router_hsts_scope comes from an environment variable.  The signature of `hstsHeaderValue` cannot contain an enum type unless I want to convert the environment variable into an enum type within the HAProxy template.  Since the HAProxy template is the only thing that calls `hstsHeaderValue`, it is not worth investigating how to change an env variable into an enum type withing the HAProxy template.  I hope that clarifies it.

Reply:
Environment variables are always passed as strings, aren't they? 

I'll I'm recommending is setting `AllHSTSHeaderPolicyScope` and `LimitedHSTSHeaderPolicyScope` to be of some new type, say `HSTSHeaderPolicyScope`, so that the new `Scope` field could be configured to only accept the strings `All` and `Limited`. How does that sound to you? That should not affect the router code changes at all, right?

Reply:
Yes, I tried this before and was doubtful about using the enum in the signature of the method. I can just pass it in as a string and cast it within the method.  I'll do that.